src/app/services/data.service.ts

```diff
@@ -15,17 +17,37 @@ import { SelectionService } from './selection.service';
 })
 export class DataService {
 
-  $dataSet = new BehaviorSubject<DataSet>(DEMO_DATA_SETS[0]);
   $aggregate = new BehaviorSubject<Aggregate|null>(null);
+  $dataSet = new BehaviorSubject<DataSet>(DEMO_DATA_SETS[0]);
+  $errors = new BehaviorSubject<string[]>([]);
+  $schema = new BehaviorSubject<JSONSchemaType<Aggregate>|null>(null);
+
+  ajv = new Ajv();
+  schemaUrl = 'https://raw.githubusercontent.com/ncsa/phyloflow/fcbf6b8380a6704f27b4049ed7ac588a25717f1e/aggregate_json/code/aggregate.schema';
 
   rootColor = '#90959B';
   clusterColors = ['#173858', '#2A5E72', '#C9C943', '#E89680', '#B26799', '#7524A1', '#489669', '#deac5b', '#bd486d'];
 
   constructor(private selectionService: SelectionService, private http: HttpClient) {
-    this.getDataSet().pipe(
-      switchMap(ds => this.http.get<Aggregate>(ds.url)),
-    ).subscribe(ds => {
-      this.$aggregate.next(ds);
+    this.http.get<JSONSchemaType<Aggregate>>(this.schemaUrl).subscribe(schema => {
+      this.$schema.next(schema);
+    });
+    combineLatest([
+      this.getDataSet(),
+      this.$schema
+    ]).pipe(
+      filter(([ds, schema]) => !!schema),
+      switchMap(([ds, schema]) => this.http.get<Aggregate>(ds.url))
+    ).subscribe({
+      next: agg => {
+        const validate = this.ajv.compile(this.$schema.value as JSONSchemaType<Aggregate>);
+        if (validate(agg)) {
+          this.$aggregate.next(agg);
+        } else {
+          this.addError('Error parsing data file: ' + validate.errors![0].message + '. Please refresh, check your file format, and try again.');
+        }
+      },
+      error: () => this.addError('The selected data file could not be parsed. Please refresh, check your file format, and try again.')
```

Comment:
bummer that the interface is forcing you to have to refresh the entire page before trying again (or at least it didn't work for me). hopefully we can change this in the future so the user can simply click "change data" and try again (no page refresh required)